### PR TITLE
REVSDL-2470 HMICommandFactory to handle GPB message from HMI

### PR DESCRIPTION
**Pull request by giangtqh**
@bonewell 
@KKarlash 
@VladSemenyuk 
@DKlimenko 
@thinhts 
please review

### CODE REVIEW

File: src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+#include "vr_cooperation/event_engine/event_observer.h"
+#include "vr_cooperation/interface/hmi.pb.h"
+
+namespace vr_cooperation {
+class VRModule;
+
+namespace commands {
+
+class RequestFromHMI: public Command,
+  public event_engine::EventObserver<vr_hmi_api::ServiceMessage, int> {
```
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h]: public event_engine::EventObserver<vr_hmi_api::ServiceMessage, vr_hmi_api::RPCName>

File: src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * @brief SendResponse allows to send response to hmi
+   */
+
+  void SendResponse();
+  /**
+   * @brief executes specific logic of children classes
+   */
+  void virtual Execute() = 0;
```
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h]: virtual void

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h [hunk 3]
```diff
@@ -37,6 +37,12 @@
 
 namespace event_engine {
 
+enum MessageType {
+  kRequest,
+  kResponse,
+  kNotification
+};
```
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h]: how about invalid and errorresponse?

File: src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+#include "vr_cooperation/commands/command.h"
+#include "vr_cooperation/event_engine/event_observer.h"
+#include "vr_cooperation/interface/hmi.pb.h"
+
+namespace vr_cooperation {
+class VRModule;
+
+namespace commands {
+
+class RequestFromHMI: public Command,
```
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h]: Please change name, something like BaseCommandGpbRequest

File: src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+   * @brief executes specific logic of children classes
+   */
+  void virtual Execute() = 0;
+
+  /**
+   * @brief Interface method that is called whenever new event received
+   */
+  void virtual OnEvent() = 0;
+
+  vr_hmi_api::ServiceMessage message_;
```
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h]: should be private

**Comment by thinhts** [comment]
Reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h (hunk 4 above)
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h]: I don't think it is necessary. :)

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h (hunk 3 above)
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h]: I think it is enough. Event is base class. Derived class should implement event_message_type according with this enumeration. Client code using  event_engine should handle message from HMI (or mobile) and create valid Event objects or don't create at all.

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h [hunk 6]
```diff
@@ -126,18 +126,14 @@ void EventDispatcher<EventMessage, EventID>::raise_event(
...
       for (; observers_[event.id()].end() != it; ++it) {
         list = it->second;
       }
     }
 
-    if ((hmi_apis::messageType::response  ==
-        event.event_message_type())
-        || (hmi_apis::messageType::error_response ==
-            event.event_message_type())) {
+    if ((MessageType::kResponse  == event.event_message_type())) {
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: Redundant parenthesis

**Comment by bonewell** [comment]
Reviewed.

**Comment by giangtqh** [comment]
@thinhts 
@bonewell 
@KKarlash 
I fixed your review notes,
@VladSemenyuk 
@DKlimenko 
please review

**Comment by bonewell** [comment]
Reviewed.

**Comment by KKarlash** [comment]
Reviewed

**Comment by DKlimenko** [comment]
reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  void SendRequest();
+
+  /**
+   * @brief SendResponse allows to send response to hmi
+   */
+
+  void SendResponse();
+  /**
+   * @brief executes specific logic of children classes
+   */
```
Comment (DKlimenko) [inline src/components/vr_cooperation/include/vr_cooperation/commands/request_from_hmi.h]: please add empty line

**Pull request merged**